File: branches/tfs-test/native-tfs/src/main/java/com/taobao/common/tfs/impl/FSName.java
/*
 * (C) 2007-2010 Alibaba Group Holding Limited.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 */
package com.taobao.common.tfs.impl;

import java.util.Arrays;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import com.taobao.common.tfs.TfsException;

public class FSName {
    public static final int MAIN_FILENAME_LENGTH = 18;
    // just for compatibility
    public static final int STANDARD_SUFFIX_LENGTH = 4;

    public static final char TFS_SMALL_KEY_CHAR = 'T';
    public static final char TFS_LARGE_KEY_CHAR = 'L';

    private static final int ENCODE_FILENAME_LENGTH = 16;
    private static final int ENCODE_FILECODE_LENGTH = 12;
    private static final int ENCODE_SIMPLECODE_LENGTH = 8;

    private char tfsClusterIndex;
    private byte[] fileCode;

    private static final byte[] tagTable = ("Taobao-inc").getBytes();
    private static final int tagLength = tagTable.length;
    private static final byte[] encodingTable =
        ("0JoU8EaN3xf19hIS2d.6pZRFBYurMDGw7K5m4CyXsbQjg_vTOAkcHVtzqWilnLPe").getBytes();
    private static final byte[] decodingTable = {
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0,0,0,0,0,0,0,0,0,0,0,0,0,18,0,0,11,16,8,36,34,19,32,4,12,0,0,0,0,0,
        0,0,49,24,37,29,5,23,30,52,14,1,33,61,28,7,48,62,42,22,15,47,3,53,57,
        39,25,21,0,0,0,0,45,0,6,41,51,17,63,10,44,13,58,43,50,59,35,60,2,20,
        56,27,40,54,26,46,31,9,38,55,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
        0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
        ,0,0,0,0,0,0,0,0,0,0,0,0};

    public FSName(String tfsName) throws TfsException {
        tfsClusterIndex = '1';
        set(tfsName);
    }

    public FSName(int clusterId, int blockId, long fileId) {
        tfsClusterIndex = (char)('0' + clusterId);
        set(blockId, fileId);
    }

    public FSName(int blockId, long fileId) {
        tfsClusterIndex = '1';
        set(blockId, fileId);
    }

    public FSName(String mainFileName, String suffix) throws TfsException {
        tfsClusterIndex = '1';
        set(mainFileName, suffix);
    }

    private int check(String mainFileName, String suffix) {
        if (mainFileName == null || mainFileName.length() == 0) {
            return 0;
        }

        int length = mainFileName.length();
        if (length < MAIN_FILENAME_LENGTH ||
            (mainFileName.charAt(0) != TFS_SMALL_KEY_CHAR &&
             mainFileName.charAt(0) != TFS_LARGE_KEY_CHAR) ||
            (mainFileName.charAt(1) < '0' || mainFileName.charAt(1) > '9')) {
            return -1;
        }

        if (length > MAIN_FILENAME_LENGTH) { // mainname has suffix
            if (null == suffix || 0 == suffix.length()) { // no specified suffix, use mainname's
                return 2;
            }
            if (! mainFileName.substring(MAIN_FILENAME_LENGTH).equals(suffix)) {
                // has specified suffix, then mainname's must equal with specified
                return -1;
            }
        }

        return 1;
    }

    private byte[] xorTagCode(byte[] input) {
        int length = input.length;
        byte[] output = new byte[length];
        for (int i = 0; i < length; i++) {
            output[i] = (byte)(input[i] ^ tagTable[i%tagLength]);
        }
        return output;
    }

    private byte[] encode(byte[] fileId, boolean simple) {
        if (fileId.length != ENCODE_FILECODE_LENGTH) {
            return null;
        }

        byte[] fileIdX = fileId;

        if (simple) {
            fileIdX = new byte[fileId.length];
            System.arraycopy(fileId, 0, fileIdX, 0, fileId.length);
            // only encode blockid & seqid, no suffix
            Arrays.fill(fileIdX, ENCODE_SIMPLECODE_LENGTH, ENCODE_FILECODE_LENGTH, (byte)0x0);
        }
        fileIdX = xorTagCode(fileIdX);

        byte[] fileName = new byte[ENCODE_FILENAME_LENGTH];
        int i, k = 0;
        int x;
        for (i = 0; i < ENCODE_FILECODE_LENGTH; i += 3) {
            x = ((fileIdX[i] << 16) & 0xff0000)
                + ((fileIdX[i + 1] << 8) & 0xff00) + (fileIdX[i + 2] & 0xff);
            fileName[k++] = encodingTable[x >>> 18];
            fileName[k++] = encodingTable[(x >>> 12) & 0x3f];
            fileName[k++] = encodingTable[(x >>> 6) & 0x3f];
            fileName[k++] = encodingTable[x & 0x3f];
        }
        return fileName;
    }

    private byte[] decode(byte[] fileName) {
        if (fileName.length != ENCODE_FILENAME_LENGTH) {
            return null;
        }

        byte[] fileId = new byte[ENCODE_FILECODE_LENGTH];

        int i, k = 0;
        int x;
        for (i = 0; i < ENCODE_FILENAME_LENGTH; i += 4) {
            x = (decodingTable[fileName[i]&0xff] << 18) +
                (decodingTable[fileName[i+1]&0xff] << 12) +
                (decodingTable[fileName[i+2]&0xff] << 6) +
                decodingTable[fileName[i+3]&0xff];
            fileId[k++] = (byte)((x >>> 16) & 0xff);
            fileId[k++] = (byte)((x >>> 8) & 0xff);
            fileId[k++] = (byte)(x & 0xff);
        }
        return xorTagCode(fileId);
    }

    private int hash(String suffix) {
        if (suffix == null) {
            return 0x80000000;
        }
        int len = suffix.length();
        int h = 0, i;
        for(i=0; i<len; i++) {
            h += suffix.charAt(i);
            h *= 7;
        }
        return (h | 0x80000000);
    }

    private void setCode(int offset, int value) {
        ByteBuffer bf;
        if (this.fileCode == null) {
            bf = ByteBuffer.allocate(ENCODE_FILECODE_LENGTH) ;
        } else {
            bf = ByteBuffer.wrap(this.fileCode);
        }
        bf.order(ByteOrder.LITTLE_ENDIAN);
        bf.putInt(offset, value);
        this.fileCode = bf.array();
    }

    private int getCode(int offset) {
        if (this.fileCode == null) return 0;
        ByteBuffer bf = ByteBuffer.wrap(this.fileCode);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        return bf.getInt(offset);
    }

    public void setBlockId(int blockId) {
        setCode(0, blockId);
    }

    public void setSeqId(int seqId) {
        setCode(4, seqId);
    }

    public void setSuffix(int suffix) {
        setCode(8, suffix);
    }

    public void setSuffix(String suffix) {
        if (suffix != null && suffix.length() > 0)
            setSuffix(hash(suffix));
    }

    public void setFileId(long fileId) {
        ByteBuffer bf = ByteBuffer.wrap(this.fileCode);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        bf.putLong(4, fileId);
        this.fileCode = bf.array();
    }

    public int getBlockId() {
        return getCode(0);
    }

    public int getSeqId() {
        return getCode(4);
    }

    public int getSuffix() {
        return getCode(8);
    }

    public long getFileId() {
        if (this.fileCode == null) return 0;
        ByteBuffer bf = ByteBuffer.wrap(this.fileCode);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        return bf.getLong(4);
    }

    private boolean set(String tfsName) throws TfsException {
        return set(tfsName, null);
    }

    private boolean set(String mainFileName, String suffix) throws TfsException {
        int ret = check(mainFileName, suffix);
        if (ret == -1) {
            throw new TfsException("illegal tfsname, name:"
                                   + mainFileName + ",suffix:" + suffix);
        } else if (ret == 0) {
            this.fileCode = new byte[ENCODE_FILECODE_LENGTH];
        } else {
            byte[] name = new byte[ENCODE_FILENAME_LENGTH];
            System.arraycopy(mainFileName.getBytes(), 2, name, 0,
                             ENCODE_FILENAME_LENGTH);
            this.fileCode = this.decode(name);

            if (ret == 2) {
                suffix = mainFileName.substring(MAIN_FILENAME_LENGTH);
            }
        }

        if (mainFileName != null && mainFileName.length() > 0) {
            tfsClusterIndex = mainFileName.charAt(1);
        }

        if (suffix != null && suffix.length() > 0) {
            setSuffix(hash(suffix));
        }

        return true;
    }

    private boolean set(int blockId, long fileId) {
        ByteBuffer bf = ByteBuffer.allocate(ENCODE_FILECODE_LENGTH);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        bf.putInt(blockId);
        bf.putLong(fileId);
        this.fileCode = bf.array();
        return true;
    }

    public String get() {
        return get(this.tfsClusterIndex, false, false);
    }

    public String get(char tfsClusterIndex) {
        return get(tfsClusterIndex, false, false);
    }

    public String get(boolean isLarge, boolean simple) {
        return get(tfsClusterIndex, isLarge, simple);
    }

    public String get(char tfsClusterIndex, boolean isLarge, boolean simple) {
        char tfsNameHead = isLarge ? TFS_LARGE_KEY_CHAR : TFS_SMALL_KEY_CHAR;
        byte[] name = new byte[MAIN_FILENAME_LENGTH];
        name[0] = (byte)(tfsNameHead & 0xff);
        name[1] = (byte)(tfsClusterIndex & 0xff);

        byte[] code = encode(this.fileCode, simple);
        System.arraycopy(code, 0, name, 2, ENCODE_FILENAME_LENGTH);
        return new String(name);
    }

    public char getTfsClusterIndex() {
        return tfsClusterIndex;
    }

    public void setTfsClusterIndex(char tfsClusterIndex) {
        this.tfsClusterIndex = tfsClusterIndex;
    }


}

File: tags/release-2.6.0/src/new_client/fsname.cpp
/*
 * (C) 2007-2010 Alibaba Group Holding Limited.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 *
 * Version: $Id: fsname.cpp 19 2010-10-18 05:48:09Z nayan@taobao.com $
 *
 * Authors:
 *   duolong <duolong@taobao.com>
 *      - initial release
 *
 */
#include <tbsys.h>
#include <string.h>

#include "encode.h"
#include "fsname.h"

using namespace tfs::common;
using namespace tbsys;
using namespace tfs::client::coding;

namespace tfs
{
  namespace client
  {
    static int32_t hash(const char *str)
    {
      if (str == NULL)
      {
        return 0;
      }
      int32_t len = strlen(str);
      int32_t h = 0;
      int32_t i = 0;
      for (i = 0; i < len; ++i)
      {
        h += str[i];
        h *= 7;
      }
      return (h | 0x80000000);
    }

    FSName::FSName() : is_valid_(true), cluster_id_(0)
    {
      file_name_[0] = '\0';
      memset(&file_, 0, sizeof(FileBits));
    }

    FSName::FSName(const uint32_t block_id, const uint64_t file_id, const int32_t cluster_id) :
      is_valid_(true), cluster_id_(cluster_id)
    {
      file_.block_id_ = block_id;
      set_file_id(file_id);
      file_name_[0] = '\0';
    }

    FSName::FSName(const char* file_name, const char* suffix, const int32_t cluster_id) : is_valid_(true)
    {
      set_name(file_name, suffix, cluster_id);
    }

    FSName::~FSName()
    {

    }

    void FSName::set_name(const char* file_name, const char* suffix, const int32_t cluster_id)
    {
      file_name_[0] = '\0';
      cluster_id_ = cluster_id;
      memset(&file_, 0, sizeof(FileBits));

      if (NULL != file_name && file_name[0] != '\0')
      {
        if (check_file_type(file_name) == INVALID_TFS_FILE_TYPE)
        {
          is_valid_ = false;
        }
        else
        {
          decode(file_name + 2, (char*) &file_);
          if (NULL == suffix)
          {
            suffix = file_name + FILE_NAME_LEN;
          }
          set_suffix(suffix);
          if (0 == cluster_id_)
          {
            cluster_id_ = file_name[1] - '0';
          }
        }
      }
    }

    const char* FSName::get_name(const bool large_flag)
    {
      if (file_name_[0] == '\0')
      {
        encode((char*) &file_, file_name_ + 2);
        if (large_flag)
        {
          file_name_[0] = LARGE_TFS_FILE_KEY_CHAR;
        }
        else
        {
          file_name_[0] = SMALL_TFS_FILE_KEY_CHAR;
        }
        file_name_[1] = static_cast<char> ('0' + cluster_id_);
        file_name_[FILE_NAME_LEN] = '\0';
      }

      return file_name_;
    }

    void FSName::set_suffix(const char *suffix)
    {
      if ((suffix != NULL) && (suffix[0] != '\0'))
      {
        file_.suffix_ = hash(suffix);
      }
    }

    string FSName::to_string()
    {
      char buffer[256];
      snprintf(buffer, 256, "block_id: %u, file_id: %"PRI64_PREFIX"u, seq_id: %u, suffix: %u, name: %s",
               file_.block_id_, get_file_id(), file_.seq_id_, file_.suffix_, get_name());
      return string(buffer);
    }

    TfsFileType FSName::check_file_type(const char* tfs_name)
    {
      TfsFileType file_type = INVALID_TFS_FILE_TYPE;
      if (NULL != tfs_name &&
          static_cast<int32_t>(strlen(tfs_name)) >= FILE_NAME_LEN)
      {
        if (LARGE_TFS_FILE_KEY_CHAR == tfs_name[0])
        {
          file_type = LARGE_TFS_FILE_TYPE;
        }
        else if (SMALL_TFS_FILE_KEY_CHAR == tfs_name[0])
        {
          file_type = SMALL_TFS_FILE_TYPE;
        }
      }
      return file_type;
    }
  }
}
